Sources/AsyncAlgorithms/Zip/Zip2StateMachine.swift

```diff
@@ -0,0 +1,367 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+struct Zip2StateMachine<Element1, Element2>: Sendable
+where Element1: Sendable, Element2: Sendable {
+  typealias SuspendedDemand = UnsafeContinuation<(Result<Element1, Error>, Result<Element2, Error>)?, Never>
+
+  private enum State {
+    case initial
+    case awaitingDemandFromConsumer(
+      task: Task<Void, Never>?,
+      suspendedBases: [UnsafeContinuation<Void, Never>]
+    )
+    case awaitingBaseResults(
+      task: Task<Void, Never>?,
+      result1: Result<Element1, Error>?,
+      result2: Result<Element2, Error>?,
+      suspendedBases: [UnsafeContinuation<Void, Never>],
+      suspendedDemand: SuspendedDemand?
+    )
+    case finished
+  }
+
+  private var state: State = .initial
+
+  mutating func taskIsStarted(
+    task: Task<Void, Never>,
+    suspendedDemand: SuspendedDemand
+  ) {
+    switch self.state {
+      case .initial:
+        self.state = .awaitingBaseResults(
+          task: task,
+          result1: nil,
+          result2: nil,
+          suspendedBases: [],
+          suspendedDemand: suspendedDemand
+        )
+
+      default:
+        preconditionFailure("Inconsistent state, the task cannot start while the state is other than initial")
+    }
+  }
+
+  enum NewDemandFromConsumerOutput {
+    case resumeBases(suspendedBases: [UnsafeContinuation<Void, Never>])
+    case startTask(suspendedDemand: SuspendedDemand)
+    case terminate(suspendedDemand: SuspendedDemand)
+  }
+
+  mutating func newDemandFromConsumer(
+    suspendedDemand: UnsafeContinuation<(Result<Element1, Error>, Result<Element2, Error>)?, Never>
+  ) -> NewDemandFromConsumerOutput {
+    switch self.state {
+      case .initial:
+        return .startTask(suspendedDemand: suspendedDemand)
+
+      case .awaitingDemandFromConsumer(let task, let suspendedBases):
+        self.state = .awaitingBaseResults(task: task, result1: nil, result2: nil, suspendedBases: [], suspendedDemand: suspendedDemand)
+        return .resumeBases(suspendedBases: suspendedBases)
+
+      case .awaitingBaseResults:
+        preconditionFailure("Inconsistent state, a demand is already suspended")
+
+      case .finished:
+        return .terminate(suspendedDemand: suspendedDemand)
+    }
+  }
+
+  enum NewLoopFromBaseOutput {
+    case none
+    case resumeBases(suspendedBases: [UnsafeContinuation<Void, Never>])
+    case terminate(suspendedBase: UnsafeContinuation<Void, Never>)
+  }
+
+  mutating func newLoopFromBase1(suspendedBase: UnsafeContinuation<Void, Never>) -> NewLoopFromBaseOutput {
```

Comment:
That’s what I’ve done in my repo « AsyncExtensions » to support variadic params but that relies on that fact that the array of elements has à consistent element type.

In our case Element1 != Element2 != Element3 … how would you do that (store an array of Any and then recast them in the good type in the AsyncZipSequence ?)

Reply:
Right, I think we can a solution for this. For the time being we can actually just use tuples to fix this. We can make my proposed `Upstream` struct generic over an element `Upstream<Element>` and then instead of storing an array we can just store a tuple for now.

```swift
 case awaitingBaseResults(
     task: Task<Void, Never>?,
     upstreams: (Upstream<Base1.Element>, Upstream<Base2.Element>, Upstream<Base3.Element>),
     suspendedDemand: SuspendedDemand?
)
```

We can also go one step further and introduce an `Upstreams` type for this. This should work for both the 2 and 3 arity case. Once we get variadic generics we probably want to create a fix sized storage that has as many slots as the length of the parameter pack.